overlord/snapstate/handlers.go

```diff
@@ -1290,13 +1369,24 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	// if we have more revisions of this snap available, then restore the
+	// revision config and save the disabled services (which are restored when
+	// we go to link another snap)
 	if len(snapst.Sequence) > 0 {
+		// save the disabled services for later
+		st.Unlock()
+		err = m.saveSnapDisabledServices(snapst, NewTaskProgressAdapterUnlocked(t))
```

Comment:
I'm not clear why we call this here,  while for config we do a restore?

Reply:
because we need to finish linking the old revision that we are going back to as part of the undo _before_ we can re-disable the services, however since we successfully finished the link snap, the services were already re-disabled and the list in snapst is now stale so we have to capture that state from systemd again before doing the unlink.

A concrete example:
1. snap is linked successfully (say as part of a refresh), previously disabled services are now disabled again and the list in snapst now only contains the services that aren't present in the revision we just linked
2. something happens later on where we need to undo the link the snap, we start executing the undoLinkSnap handler
3. we need to save the list of disabled services again since we transferred that info to systemd, so we do it now, saving to snapst
4. we unlink the revision, and delete the service files from systemd
5. we link the old version of the snap, install the old snap service files
6. after the old service files are available we look at the disabled services in snapst and restore those being disabled.

Reply:
because we need to finish linking the old revision that we are going back to as part of the undo before we can re-disable the services, however since we successfully finished the link snap, the services were already re-disabled and the list in snapst is now stale so we have to capture that state from systemd again before doing the unlink.

A concrete example:

snap is linked successfully (say as part of a refresh), previously disabled services are now disabled again and the list in snapst now only contains the services that aren't present in the revision we just linked
something happens later on where we need to undo the link the snap, we start executing the undoLinkSnap handler
we need to save the list of disabled services again since we transferred that info to systemd, so we do it now, saving to snapst
we unlink the revision, and delete the service files from systemd
we link the old version of the snap, install the old snap service files
after the old service files are available we look at the disabled services in snapst and restore those being disabled.

Reply:
I'm very worried about how long the explanation for any of this lines are. We need to simplify, make all of this more ovious/symmetric,  moving all interaction with systemd vs state manip to stop/start-service is even more tempting.

Reply:
sorry, here I got confused by the context of the nearby config code (which wer are now trying to understand whether it belongs here)